src/main/java/com/google/finapp/SpannerDaoJDBCImpl.java

```diff
@@ -0,0 +1,191 @@
+      throws SpannerDaoException {
+    try {
+      try (Connection connection = DriverManager.getConnection(this.connectionUrl)) {
+        try (PreparedStatement ps =
+            connection.prepareStatement(
+                "INSERT INTO Customer\n"
+                    + "(CustomerId, Name, Address)\n"
+                    + "VALUES\n"
+                    + "(?, ?, ?)")) {
+          ps.setBytes(1, customerId.toByteArray());
+          ps.setString(2, name);
+          ps.setString(3, address);
+          ps.executeUpdate();
+        }
+      }
+    } catch (SQLException e) {
+      throw new SpannerDaoException(e);
+    }
+  }
+
+  public void createAccount(
+      ByteArray accountId, AccountType accountType, AccountStatus accountStatus, BigDecimal balance)
+      throws SpannerDaoException {
+    try {
+      try (Connection connection = DriverManager.getConnection(this.connectionUrl)) {
+        try (PreparedStatement ps =
+            connection.prepareStatement(
+                "INSERT INTO Account\n"
+                    + "(AccountId, AccountType, AccountStatus, Balance, CreationTimestamp)\n"
+                    + "VALUES\n"
+                    + "(?, ?, ?, ?, PENDING_COMMIT_TIMESTAMP())")) {
+          ps.setBytes(1, accountId.toByteArray());
+          ps.setInt(2, accountType.getNumber());
+          ps.setInt(3, accountStatus.getNumber());
+          ps.setBigDecimal(4, balance);
+          ps.executeUpdate();
+        }
+      }
+    } catch (SQLException e) {
+      throw new SpannerDaoException(e);
+    }
+  }
+
+  public void addAccountForCustomer(
+      ByteArray customerId, ByteArray accountId, ByteArray roleId, String roleName)
+      throws SpannerDaoException {
+    try {
+      try (Connection connection = DriverManager.getConnection(this.connectionUrl)) {
+        try (PreparedStatement ps =
+            connection.prepareStatement(
+                "INSERT INTO CustomerRole\n"
+                    + "(CustomerId, AccountId, RoleId, Role)\n"
+                    + "VALUES\n"
+                    + "(?, ?, ?, ?)")) {
+          ps.setBytes(1, customerId.toByteArray());
+          ps.setBytes(2, accountId.toByteArray());
+          ps.setBytes(3, roleId.toByteArray());
+          ps.setString(4, roleName);
+          ps.executeUpdate();
+        }
+      }
+    } catch (SQLException e) {
+      throw new SpannerDaoException(e);
+    }
+  }
+
+  public void moveAccountBalance(ByteArray fromAccountId, ByteArray toAccountId, BigDecimal amount)
+      throws SpannerDaoException {
+    try {
+      try (Connection connection = DriverManager.getConnection(this.connectionUrl)) {
+        connection.setAutoCommit(false);
+        byte[] fromAccountIdArray = fromAccountId.toByteArray();
+        byte[] toAccountIdArray = toAccountId.toByteArray();
+        BigDecimal[] accountBalances =
+            readAccountBalances(fromAccountIdArray, toAccountIdArray, connection);
```

Comment:
Do you think you might use readAccountBalances anywhere else? If not, have you considered inlining the logic here?

This would allow you to define separate:
BigDecimal sourceAmount;
BigDecimal destAmount;
and avoid having to deal with BigDecimal[] and indexes that are not as readable.

What do you think?

Reply:
Good idea. I mainly did this separate function to abstract away database interactions into separate helper functions (like `updateAccount` and `insertTransaction`) to make the logic clearer, rather than for the sake of reuse. However, at this point, this helper function is being more confusing than helpful, so I'll move it inline.